src/backend/metal/src/device.rs

```diff
@@ -1386,69 +1383,117 @@ impl hal::device::Device<Backend> for Device {
+        data: Option<&[u8]>,
     ) -> Result<n::PipelineCache, d::OutOfMemory> {
-        Ok(n::PipelineCache {
-            #[cfg(feature = "cross")]
-            modules: FastStorageMap::default(),
-        })
+        let device = self.shared.device.lock();
+
+        let create_binary_archive = |data: &[u8]| {
+            let supports_binary_archive = device.supports_family(metal::MTLGPUFamily::Apple3)
+                || device.supports_family(metal::MTLGPUFamily::Mac1);
+
+            if supports_binary_archive {
+                let descriptor = metal::BinaryArchiveDescriptor::new();
+
+                // We need to keep the temp file alive so that it doesn't get deleted until after a
+                // binary archive has been created.
+                let _temp_file = if !data.is_empty() {
+                    // It would be nice to use a `data:text/plain;base64` url here and just pass in a
+                    // base64-encoded version of the data, but metal validation doesn't like that:
+                    // -[MTLDebugDevice newBinaryArchiveWithDescriptor:error:]:1046: failed assertion `url, if not nil, must be a file URL.'
+
+                    let temp_file = tempfile::NamedTempFile::new().unwrap();
+
+                    temp_file.as_file().write_all(&data).unwrap();
+
+                    let url = metal::URL::new_with_string(&format!(
+                        "file://{}",
+                        temp_file.path().display()
+                    ));
+
+                    descriptor.set_url(&url);
+
+                    Some(temp_file)
+                } else {
+                    None
+                };
+
+                Ok(Some(n::BinaryArchive {
+                    inner: device
+                        .new_binary_archive_with_descriptor(&descriptor)
+                        .map_err(|_| d::OutOfMemory::Device)?,
+                    is_empty: AtomicBool::new(data.is_empty()),
+                }))
+            } else {
+                Ok(None)
+            }
+        };
+
+        if let Some(data) = data.filter(|data| !data.is_empty()) {
+            let pipeline_cache: n::SerializablePipelineCache = bincode::deserialize(data).unwrap();
+
+            Ok(n::PipelineCache {
+                binary_archive: create_binary_archive(&pipeline_cache.binary_archive)?,
+                spv_to_msl: n::load_spv_to_msl_cache(pipeline_cache.spv_to_msl),
+            })
+        } else {
+            Ok(n::PipelineCache {
+                binary_archive: create_binary_archive(&[])?,
+                spv_to_msl: Default::default(),
+            })
+        }
     }
 
     unsafe fn get_pipeline_cache_data(
         &self,
-        _cache: &n::PipelineCache,
+        cache: &n::PipelineCache,
     ) -> Result<Vec<u8>, d::OutOfMemory> {
-        //empty
-        Ok(Vec::new())
+        let binary_archive = || {
+            let binary_archive = match &cache.binary_archive {
+                Some(binary_archive) => binary_archive,
```

Comment:
our clippy is configured to not like matching on references. Instead, let's be explicit here:
```rust
match cache.binary_archive {
  Some(ref archive) => archive,
   ...
}
```